vsphere_plugin_common/__init__.py

```diff
@@ -596,139 +765,527 @@ def get_server_list(self):
         )
-        for datastore in datastore_list:
-            if datastore._moId not in except_datastores:
-                dtstr_free_spc = datastore.info.freeSpace
-                if selected_datastore is None:
-                    selected_datastore = datastore
-                    ctx.logger.debug(
-                        'Using first datastore {name} with remaining '
-                        '{space}'.format(
-                            name=datastore.name,
-                            space=datastore.info.freeSpace,
-                        )
+
+        if allowed_hosts:
+            hosts = [host for host in hosts if host.name in allowed_hosts]
+            ctx.logger.debug(
+                'Filtered list of hosts to be considered: {hosts}'.format(
+                    hosts=', '.join(host_names),
+                )
+            )
+
+        if allowed_clusters:
+            cluster_list = self.get_clusters()
+            cluster_names = [cluster.name for cluster in cluster_list]
+            valid_clusters = set(allowed_clusters).union(set(cluster_names))
+            ctx.logger.debug(
+                'Only hosts on the following clusters will be used: '
+                '{clusters}'.format(
+                    clusters=', '.join(valid_clusters),
+                )
+            )
+
+        candidate_hosts = []
+        for host in hosts:
+            if not self.host_is_usable(host):
+                ctx.logger.warn(
+                    'Host {host} not usable due to health status.'.format(
+                        host=host.name,
                     )
-                else:
-                    selected_dtstr_free_spc = selected_datastore.info.freeSpace
-                    ctx.logger.debug(
-                        'Checking datastore {name} with remaining '
-                        '{space}'.format(
-                            name=datastore.name,
-                            space=datastore.info.freeSpace,
+                )
+                continue
+
+            if allowed_clusters:
+                cluster = self.get_host_cluster_membership(host)
+                if cluster not in allowed_clusters:
+                    if cluster:
+                        ctx.logger.warn(
+                            'Host {host} is in cluster {cluster}, '
+                            'which is not an allowed cluster.'.format(
+                                host=host.name,
+                                cluster=cluster,
+                            )
+                        )
+                    else:
+                        ctx.logger.warn(
+                            'Host {host} is not in a cluster, '
+                            'and allowed clusters have been set.'.format(
+                                host=host.name,
+                            )
                         )
+                    continue
+
+            free_memory = self.get_host_free_memory(host)
+
+            if free_memory - vm_memory < 0:
```

Comment:
free_memory is available memory on hypervisor
vm_memory is memory assigned to vm

Therefore, free memory - vm_memory needs to be positive or we refuse to deploy

(we could actually allow it due to ballooning, etc, but I'd rather wait for a feature request than deal with the complexity of that logic currently)